File: 673.numberOfLongestIncreasingSubsequence/numberOfLongestIncreasingSubsequence.cpp
// clang++ numberOfLongestIncreasingSubsequence.cpp -std=c++11

#include <iostream>
#include <vector>
using namespace std;

class Solution{
public:
  int findNumberOfLIS(vector<int>& nums) {
    int res = 0;
    int size = nums.size();
    if (size == 0)
      return res;
    vector<pair<int, int> > dp(size, {1,1}); // first: lens, second: cnts
    int maxLen = 1;
    for (int i = 1; i < size; i++) {
      for (int j = 0; j < i; j++) {
	if (nums[j] < nums[i]) {
	  if (dp[j].first + 1 == dp[i].first) {
	    dp[i].second += dp[j].second; // equals dp[j].second + 1 because dp[i].second == 1
	  }
	  if (dp[j].first + 1 > dp[i].first) {
	    dp[i].first = dp[j].first + 1;
	    dp[i].second = dp[j].second; // no need to add one, just update
	  }
	  maxLen = max(maxLen, dp[i].first);
	}
      }
    }
    for (pair<int, int>& p : dp) {
      if (p.first == maxLen)
	res += p.second;
    }
    return res;
  }
};

int main()
{
  Solution s;
  vector<int> nums = {1,3,5,4,7};
  cout << s.findNumberOfLIS(nums) << endl;
  return 0;
}

File: 673.numberOfLongestIncreasingSubsequence/numberOfLongestIncreasingSubsequence.java
class Solution {
    public int findNumberOfLIS(int[] nums) {
        int[][] dp = new int[nums.length][2]; // 0 -> longest len, 1 -> path number
        dp[0][0] = dp[0][1] = 1;
        int longestLen = 1;
        for (int i = 1; i < nums.length; i++) {
            dp[i][0] = dp[i][1] = 1;
            for (int j = 0; j < i; j++) {
                if (nums[i] > nums[j]) {
                    if (dp[i][0] < dp[j][0] + 1) {
                        dp[i][0] = dp[j][0] + 1;
                        dp[i][1] = dp[j][1];
                    } else if (dp[i][0] == dp[j][0] + 1) {
                        dp[i][1] += dp[j][1];
                    }
                }
            }
            if (dp[i][0] > longestLen) {
                longestLen = dp[i][0];
            }
        }
        int res = 0;
        for (int i = 0; i < dp.length; i++) {
            if (dp[i][0] == longestLen) res += dp[i][1];
        }
        return res;
    }
}
